Feat/similarity

melisandequentzel opened the pull request:
Add snippet_similarity.py + add update_similar_snippets method to client_sqlite.py + update requirements.txt

marcorosa commented:
Hi @melisandequentzel thanks for the PR
Please change the base branch with `develop` instead of `main`

Then please follow all the indications of @pep8speaks that you see above to fix the style of the code. Then I will go into all the changes 😄

melisandequentzel commented:
Hi @marcorosa thanks for the remarks, will make the changes to the style shortly :)

melisandequentzel commented:
@marcorosa done, the last one: Line 2:1: F401 'tensorflow_text' imported but unused 
needs to remain because without that import some errors are raised

marcorosa commented on requirements.txt:
```diff
@@ -14,3 +14,6 @@ requests
 scikit-learn
 srsly
 tqdm
+tensorflow==2.4.1
```
please add the dependencies in alphabetical order

marcorosa commented on credentialdigger/snippet_similarity.py:
```diff
@@ -0,0 +1,38 @@
+import tensorflow as tf
+import tensorflow_text
+import logging
+import tensorflow_hub as hub
+import tensorflow.keras.preprocessing.text
+import numpy as np
```
The order of the imports is wrong (you can use [`isort`](https://pycqa.github.io/isort/) to do it automatically)

marcorosa commented on credentialdigger/snippet_similarity.py:
```diff
@@ -0,0 +1,38 @@
+logging.getLogger('tensorflow').setLevel(logging.ERROR)
+
+
+def build_embedding_model():
```
missing documentation

marcorosa commented on credentialdigger/snippet_similarity.py:
```diff
@@ -0,0 +1,38 @@
+    return snippet_embedding
+
+
+def compute_similarity(emb_1, emb_2):
```
same as above

marcorosa commented on credentialdigger/client.py:
```diff
@@ -1006,3 +1010,25 @@ def _get_scan_rules(self, category=None, exclude=None):
+                                target_snippet,
+                                state,
+                                repo_url,
+                                file_name=None,
```
why do you need the filename? As far as I understood, given a `target_snippet` you need to get all the "new" discoveries for a certain repo, and then compute each similarity and eventually flag other FPs. So what's the need of having the file name in this process?

melisandequentzel replied:
The purpose of having a filename here is to allow for exclusive updating of similar snippets found in the same file as the target snippet, because there could be a situation where a file contains similar snippets as another file, but one with real passwords and the other with dummy ones. This parameter is provided 'just in case' someone needs this feature, but is not meant to be used as a general rule. I can remove it if you think it should not be there.

marcorosa replied:
Let's keep it with this PR then, but before merging into the main branch I would like to test it. Moreover, I'm not fully convinced this parameter is really needed: if the same snippet appears in 2 different files, then either they are treated at the same way (i.e., both TP or both FP), or it is the path model that should flag one of the two as a FP.

melisandequentzel replied:
Yes in theory, but then why need a snippet model if the path model is always right? Some snippets with fake credentials can appear in paths that aren't dedicated to testing and/or detected as dummy by the path model, since we're applying a model to distinguish real from fake passwords after the path model. The snippet model may fail to recognize a fake password as being fake, and that's the whole point of having a similarity model: to provide a simple, 'user-powered' way of compensating for failings/shortcomings of one of the two models. This may become superfluous when the snippet model reaches a very high accuracy, but seems quite relevant as of now. Adding the filename parameter offers a more granular control over the update, which can't be a bad thing. In the UI, it could just take the form of a button to tick 'restrict updates to current file'. I noticed, there are often similar snippets in the same file, so it would still be efficient using it that way.

melisandequentzel replied:
(There clearly is a (very) small danger bound to the similarity feature, and steps to mitigate this danger should be taken, I think. Otherwise, we'll need to clearly inform the user on the potential risks of overly using this functionality on snippets that do contain a password).

marcorosa commented on ui/backend/client_ui.py:
```diff
@@ -97,3 +101,26 @@ def check_repo(self, repo_url, git_token=None, local_repo=False):
+                                repo_url,
+                                file_name=None,
+                                threshold=0.96):
+        discoveries = self.get_discoveries(repo_url,
+                                           file_name,
+                                           state_filter='new')[1]
```
why do you only keep the [1] instead of the whole list?
are you sure you don't raise an IndexError if you don't have other discoveries?

melisandequentzel replied:
I added [1] because the get_discoveries method from the UI classes returns the number of discoveries as first element, which is useless in this case. I'm only retriving the actual discoveries, not their count. This is what adding [1] allows for.

marcorosa commented on ui/backend/client_ui_sqlite.py:
```diff
@@ -87,6 +91,9 @@ def get_discoveries(self, repo_url, file_name=None, state_filter=None,
         # Build outer query to get all occurrences of the paginated snippets
         query = 'SELECT * FROM discoveries WHERE repo_url=?'
         params = [repo_url]
+        params = [repo_url]
+        params = [repo_url]
+        params = [repo_url]
```
?

marcorosa commented on ui/backend/client_ui_sqlite.py:
```diff
@@ -165,3 +172,26 @@ def get_files_summary(self, repo_url):
                 " GROUP BY file_name"
             ))
+
+    def update_similar_snippets(self,
```
why do you need to redefine this (or to copy it from the other class)?

melisandequentzel replied:
Precisely because of the get_discoveries method that is specific to the UI classes. No other way to go than to rewrite the adapted version of update_similar_snippets in the UI classes

melisandequentzel replied:
Or maybe checking for length of method output directly in client.py, but doesn't seem any more efficient

marcorosa commented:
Hi, apart from all the changes/doubts requested above, I have 2 further concerns?
- if/when the PR is ready to be reviewed, remember to turn it to "ready for review" and don't leave it draft
- There are few comments, and the code is hard to follow. Please write comments in the code, and remember to write the doctrings for ALL your new methods ;)

melisandequentzel commented on requirements.txt:
```diff
@@ -14,3 +14,6 @@ requests
 tqdm
+tensorflow==2.4.1
+tensorflow-text==2.4.1
+tf-models-official
```
Version 2.4.1 isn't available for tf-models-official, the right version is picked up automatically. The rule to have matching versions only applies to tf and tf-text

melisandequentzel commented:
Hi, okay, thank you for the reviews. I will add comments + docstrings, and will make sure to do so and to have the right order for imports in the future :)

marcorosa commented on ui/backend/client_ui_sqlite.py:
```diff
@@ -165,3 +169,26 @@ def get_files_summary(self, repo_url):
+                                state,
+                                repo_url,
+                                file_name=None,
+                                threshold=0.95):
```
it was 0.96 in the parent class. Why have you changed it here?

melisandequentzel replied:
Indeed, I forgot to update the value here, I thought I had done so. It should be 0.96

melisandequentzel replied:
Oh but the code should have been removed from that class altogether, that's why I didn't update it. I'm removing it now

marcorosa commented on credentialdigger/client.py:
```diff
@@ -1006,3 +1008,51 @@ def _get_scan_rules(self, category=None, exclude=None):
+                                                snippet_embedding)
+                if similarity > threshold:
+                    n_updated_snippets += 1
+                    self.update_discovery(d['id'], state)
```
does it work without passing the query like in the update_discovery method?

melisandequentzel replied:
Yes, because the update_discovery method that gets called is that of the child class, which doesn't require a query parameter. It definitely works.

marcorosa reviewed (changes requested):
You are missing the client_ui_postgres class

melisandequentzel commented on credentialdigger/client.py:
```diff
@@ -1006,3 +1008,51 @@ def _get_scan_rules(self, category=None, exclude=None):
+            restrict to a given file the search for similar snippets
+        threshold: double
+            update snippets with similarity score above threshold.
+            Values lesser than 0.94 do not generally imply any relevant
```
0.94, it's intended as a guideline on which values to use for this parameter. 0.96 is usually perfect, but in some rare cases, putting a lower value will ensure all similar snippets will be detected; however that value should always be greater than 0.94.

melisandequentzel commented on ui/backend/client_ui.py:
```diff
@@ -97,3 +101,49 @@ def check_repo(self, repo_url, git_token=None, local_repo=False):
+            restrict to a given file the search for similar snippets
+        threshold: double
+            update snippets with similarity score above threshold.
+            Values lesser than 0.94 do not generally imply any relevant
```
0.94

melisandequentzel commented:
@marcorosa I'm not missing the code for the client_ui_postgres file because it's in the parent class (client_ui file). I had failed to remove the code from the client_ui _sqlite file. I just did so, and re-tested it.

melisandequentzel commented:
@marcorosa I fixed the float types and removed code from client_ui_sqlite, so the code should be ready by now. I replied to your comment regarding the filename parameter in the update_similar_snippets method. Should I make the change anyway?
